### PR TITLE
[10.0] add new module to manage sale promotion based on rules

### PR DESCRIPTION
**Pull request by lmignon**
A new repository 'sale_discount' could be a better place for this kind of addon.

### CODE REVIEW

File: sale_promotion_rule/__manifest__.py [hunk 1]
```diff
@@ -0,0 +1,33 @@
...
+# Copyright 2017 Akretion (http://www.akretion.com)
+# Benoît GUILLOT <benoit.guillot@akretion.com>
+# License AGPL-3.0 or later (http://www.gnu.org/licenses/agpl).
+{
+    "name": "Sale Promotion Rule",
+    "summary": "Module to manage promotion rule on sale order",
+    "version": "10.0.1.0.0",
+    "category": "Sale",
+    "website": "https://akretion.com",
+    "author": "Akretion",
```
Comment (sebastienbeau) [inline sale_promotion_rule/__manifest__.py]: Missing author "Acsone, Odoo Community Association (OCA)"
```suggestion
    "author": "Akretion, Acsone, Odoo Community Association (OCA)",
```

File: sale_promotion_rule/README.rst [hunk 2]
```diff
@@ -0,0 +1,44 @@
...
+* Benoît GUILLOT <benoit.guillot@akretion.com>
+
+
+Roadmap
+----------
+
+* Add "automatic" rule that are automatically applied without coupon (only coupon have been implmented)
+* Add gift support. (Only discount offer have been implemented)
+* Add discount amount support
+* Improve condition by using custom filter on partner
```
Comment (sebastienbeau) [inline sale_promotion_rule/README.rst]: Missing OCA README info

**Review by sebastienbeau — APPROVED** [review]
Fonctionnal tested
Thanks for proposing in OCA.
Minor change to do regarding author and readme

File: sale_promotion_rule/README.rst [hunk 3]
```diff
@@ -0,0 +1,44 @@
...
+.. image:: static/description/promotion_rule.png
+   :alt: .
+
+
+Contributors
+------------
+
+* BEAU Sébastien <sebastien.beau@akretion.com>
+* Benoît GUILLOT <benoit.guillot@akretion.com>
+
```
Comment (sebastienbeau) [inline sale_promotion_rule/README.rst:86]: you can add yourself (Laurent) as contributor

Comment (lmignon) [reply sale_promotion_rule/README.rst:86]: Thank you @sebastienbeau. Comments included in last commit.

**Review by Cedric-Pigeon — APPROVED** [review]

File: sale_promotion_rule/readme/DESCRIPTION.rst [hunk 4]
```diff
@@ -0,0 +1,18 @@
...
+Automatic rules are applied/recomputed automatically for a sale order when the
+user clicks on the button "Apply discount" in the view form. Depending on the
+rule's criteria more than one automatic rule can be applied to a same sale
+order.
+
+Coupon are special manual rules. Only one coupon can be applied to a
+sale order. This rule takes always precedence over automatic rules.
+
+
+.. image:: https://raw.githubusercontent.com/OCA/sale-workflow/10.0/sale_promotion_rule/static/description/promotion_rule.png
```
Comment (legalsylvain) [inline sale_promotion_rule/readme/DESCRIPTION.rst]: You can now write : ../sale_promotion_rule/static/description/promotion_rule.png

(more simple, and make the url resilient, during the migration)

Not a blocking point.

Comment (lmignon) [reply sale_promotion_rule/readme/DESCRIPTION.rst]: thank you @legalsylvain Change done

File: sale_promotion_rule/__manifest__.py [hunk 5]
```diff
@@ -0,0 +1,34 @@
...
+# Benoît GUILLOT <benoit.guillot@akretion.com>
+# License AGPL-3.0 or later (http://www.gnu.org/licenses/agpl).
+{
+    "name": "Sale Promotion Rule",
+    "summary": "Module to manage promotion rule on sale order",
+    "version": "10.0.1.0.0",
+    "development_status": "Beta",
+    "category": "Sale",
+    "website": "https://akretion.com",
+    'author': 'Anybox, ACSONE SA/NV,Odoo Community Association (OCA)',
```
Comment (rvalyi) [inline sale_promotion_rule/__manifest__.py]: this is strange: we have akretion in the website but not in the author

Comment (lmignon) [reply sale_promotion_rule/__manifest__.py]: I'll fix it... Should be Akretion in place of Anybox...

Comment (lmignon) [reply sale_promotion_rule/__manifest__.py]: @rvalyi fixed in https://github.com/OCA/sale-workflow/pull/729/commits/6fb82acd4c3538132d00a843acb58fe92afbb803

File: sale_promotion_rule/__manifest__.py [hunk 6]
```diff
@@ -0,0 +1,34 @@
...
+# Copyright 2017 Akretion (http://www.akretion.com)
+# Benoît GUILLOT <benoit.guillot@akretion.com>
+# License AGPL-3.0 or later (http://www.gnu.org/licenses/agpl).
+{
+    "name": "Sale Promotion Rule",
+    "summary": "Module to manage promotion rule on sale order",
+    "version": "10.0.1.0.0",
+    "development_status": "Beta",
+    "category": "Sale",
+    "website": "https://akretion.com",
```
Comment (nikul-serpentcs) [inline sale_promotion_rule/__manifest__.py]: Replace URL as per [this](https://github.com/OCA/maintainer-tools/blob/master/template/module/__manifest__.py#L11)

File: sale_promotion_rule/__manifest__.py [hunk 7]
```diff
@@ -0,0 +1,34 @@
...
+    'author': 'Akretion, ACSONE SA/NV, Odoo Community Association (OCA)',
+    "license": "AGPL-3",
+    "application": False,
+    "installable": True,
+    "external_dependencies": {
```
Comment (nikul-serpentcs) [inline sale_promotion_rule/__manifest__.py]: @lmignon if not any `external_dependencies` then remove it

**Review by nikul-serpentcs — COMMENTED** [review]
Improve Code

File: sale_promotion_rule/__manifest__.py (hunk 6 above)
Comment (lmignon) [reply sale_promotion_rule/__manifest__.py]: @nikul-serpentcs done

File: sale_promotion_rule/__manifest__.py (hunk 7 above)
Comment (lmignon) [reply sale_promotion_rule/__manifest__.py]: @nikul-serpentcs done

File: sale_promotion_rule/models/sale_order.py [hunk 8]
```diff
@@ -0,0 +1,64 @@
...
+        for rec in self:
+            rec.has_promotion_rules = (
+                rec.coupon_promotion_rule_id or
+                rec.promotion_rule_ids)
+
+    @api.depends('promotion_rule_ids', 'coupon_promotion_rule_id')
+    def _compute_applied_promotion_rule_ids(self):
+        for rec in self:
+            rec.applied_promotion_rule_ids = (
+                rec.coupon_promotion_rule_id + rec.promotion_rule_ids)
```
Comment (rousseldenis) [inline sale_promotion_rule/models/sale_order.py:52]: Even if each rule is cumulated, IMHO use of union is better

File: sale_promotion_rule/models/sale_promotion_rule.py [hunk 9]
```diff
@@ -0,0 +1,317 @@
...
+        column1='rule_id',
+        column2='pricelist_id',
+        string='Restricted pricelists')
+    usage_restriction = fields.Selection(
+        selection=[
+            ('one_per_partner', 'One per partner'),
+            ('no_restriction', 'No restriction')
+        ], default='no_restriction',
+        required=True)
+    restriction_amount = fields.Selection(
```
Comment (rousseldenis) [inline sale_promotion_rule/models/sale_promotion_rule.py]: IMHO the name is quite weird as we are waiting for a float. Maybe 'restriction_amount_filed ?

File: sale_promotion_rule/models/sale_promotion_rule.py [hunk 10]
```diff
@@ -0,0 +1,317 @@
...
+    def _check_valid_total_amount(self, order):
+        precision = self.env['decimal.precision'].precision_get('Discount')
+        return float_compare(
+            self.minimal_amount,
+            order[self.restriction_amount],
+            precision_digits=precision) < 0
+
+    def _check_valid_usage(self, order):
+        if self.usage_restriction == 'one_per_partner':
+            return not self.env['sale.order'].search([
```
Comment (rousseldenis) [inline sale_promotion_rule/models/sale_promotion_rule.py]: Maybe a search_count is sufficent ?

File: sale_promotion_rule/models/sale_promotion_rule.py [hunk 11]
```diff
@@ -0,0 +1,317 @@
...
+        required=True)
+    restriction_amount = fields.Selection(
+        selection=[
+            ('amount_total', 'Taxed amount'),
+            ('amount_untaxed', 'Untaxed amount'),
+        ], default='amount_total',
+        required=True)
+    minimal_amount = fields.Float(
+        digits=dp.get_precision('Discount'))
+    display_name = fields.Char(
```
Comment (rousseldenis) [inline sale_promotion_rule/models/sale_promotion_rule.py]: You should override name_get as display_name is automatically computed on it.

File: sale_promotion_rule/models/sale_promotion_rule.py [hunk 12]
```diff
@@ -0,0 +1,317 @@
...
+    def _apply_discount_to_order_lines(self, lines):
+        self.ensure_one()
+        if not self.promo_type == 'discount':
+            return
+        if not self.discount_type == 'percentage':
+            raise ValidationError(
+                _('Discount promotion of type %s is not supported') %
+                self.discount_type)
+
+        lines_by_order = defaultdict(self.env['sale.order.line'].browse)
```
Comment (rousseldenis) [inline sale_promotion_rule/models/sale_promotion_rule.py:297]: browse() ?

**Review by rousseldenis — CHANGES_REQUESTED** [review]
LGTM. Some comments

File: sale_promotion_rule/models/sale_order.py (hunk 8 above)
Comment (lmignon) [reply sale_promotion_rule/models/sale_order.py:52]: done

File: sale_promotion_rule/models/sale_promotion_rule.py (hunk 9 above)
Comment (lmignon) [reply sale_promotion_rule/models/sale_promotion_rule.py]: done

File: sale_promotion_rule/models/sale_promotion_rule.py (hunk 10 above)
Comment (lmignon) [reply sale_promotion_rule/models/sale_promotion_rule.py]: done

File: sale_promotion_rule/models/sale_promotion_rule.py (hunk 11 above)
Comment (lmignon) [reply sale_promotion_rule/models/sale_promotion_rule.py]: done

File: sale_promotion_rule/models/sale_promotion_rule.py (hunk 12 above)
Comment (lmignon) [reply sale_promotion_rule/models/sale_promotion_rule.py:297]: @rousseldenis arg of defaultdict must be a callable....

**Review by rousseldenis — APPROVED** [review]

**Comment by pedrobaeza** [comment]
@lmignon may you squash a bit commit history before merging?

**Comment by lmignon** [comment]
@pedrobaeza Sorry for the late reply, I didn't see your comment. It's a lot of work to rework/squash the commit history since I've 3 others PRs based on this one... 
https://github.com/OCA/sale-workflow/pulls/lmignon

**Comment by pedrobaeza** [comment]
Well, tell me when you finish with all, and I'll merge this with a bit of squashing.

**Comment by lmignon** [comment]
@pedrobaeza I've reworked the commit history. This PR is ready to merge. I'll rebase the others based on this one..

**Pull request merged**